Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/storage/TestCertificateStorage.kt

```diff
@@ -0,0 +1,102 @@
+package de.rki.coronawarnapp.coronatest.storage
+
+import android.content.Context
+import androidx.core.content.edit
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import de.rki.coronawarnapp.coronatest.server.CoronaTestResult
+import de.rki.coronawarnapp.coronatest.type.CoronaTest
+import de.rki.coronawarnapp.coronatest.type.TestCertificateContainer
+import de.rki.coronawarnapp.coronatest.type.pcr.PCRCertificateContainer
+import de.rki.coronawarnapp.coronatest.type.rapidantigen.RACertificateContainer
+import de.rki.coronawarnapp.util.di.AppContext
+import de.rki.coronawarnapp.util.serialization.BaseGson
+import de.rki.coronawarnapp.vaccination.core.repository.storage.ContainerPostProcessor
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class TestCertificateStorage @Inject constructor(
+    @AppContext val context: Context,
+    @BaseGson val baseGson: Gson,
+    private val containerPostProcessor: ContainerPostProcessor,
+) {
+
+    private val prefs by lazy {
+        context.getSharedPreferences("coronatest_certificate_localdata", Context.MODE_PRIVATE)
+    }
+
+    private val gson by lazy {
+        baseGson.newBuilder().apply {
+            registerTypeAdapter(CoronaTestResult::class.java, CoronaTestResult.GsonAdapter())
+            registerTypeAdapterFactory(containerPostProcessor)
+        }.create()
+    }
+
+    private val typeTokenPCR by lazy {
+        object : TypeToken<Set<PCRCertificateContainer>>() {}.type
+    }
+
+    private val typeTokenRA by lazy {
+        object : TypeToken<Set<RACertificateContainer>>() {}.type
+    }
+
+    var testCertificates: Collection<TestCertificateContainer>
+        get() {
+            Timber.tag(TAG).d("load()")
+
+            val pcrCerts: Set<PCRCertificateContainer> = run {
+                val raw = prefs.getString(PKEY_DATA_PCR, null) ?: return@run emptySet()
+                gson.fromJson<Set<PCRCertificateContainer>>(raw, typeTokenPCR).onEach {
```

Comment:
GsonAdapter should be doing type checks ?

Reply:
I think we would need to use `RuntimeTypeAdapterFactory` for this and for some reason I decided against it also in `CoronaTestStorage` :thinking:, likely because splitting the types seemed less fragile than using the `RuntimeTypeAdapterFactory`.

Reply:
If you see a way, we could discuss it on Monday.